DynamicConfigSmokeTest is flaky

## Test method
<!--
`com.example.javaagent.smoketest.DynamicConfigSmokeTest`
-->
See for example this [GH action run](https://github.com/elastic/elastic-otel-java/pull/497/checks?check_run_id=35870442792).

## Test logs

<details>
<summary>Click to expand</summary>

```
flipSending()
	java.lang.AssertionError: java.lang.AssertionError: 
	Expected size: 0 but was: 2 in:
	[trace_id: "\225\317\211\330\311s\222\365\022\022`\f|\244\257a"
	span_id: "\212\352\r\017\'\3645\337"
	parent_span_id: "\227$\265\360f\217\344\377"
	name: "DynamicConfigController.flipSending"
	kind: SPAN_KIND_INTERNAL
	start_time_unix_nano: 1737371954187501184
	end_time_unix_nano: 1737371954187559343
	attributes {
	  key: "code.namespace"
	  value {
	    string_value: "co.elastic.otel.test.DynamicConfigController"
	  }
	}
	attributes {
	  key: "thread.id"
	  value {
	    int_value: 36
	  }
	}
	attributes {
	  key: "code.function"
	  value {
	    string_value: "flipSending"
	  }
	}
	attributes {
	  key: "thread.name"
	  value {
	    string_value: "http-nio-8080-exec-3"
	  }
	}
	status {
	}
	,
	    trace_id: "\225\317\211\330\311s\222\365\022\022`\f|\244\257a"
	span_id: "\227$\265\360f\217\344\377"
	name: "GET /dynamicconfig/flipSending"
	kind: SPAN_KIND_SERVER
	start_time_unix_nano: 1737371954186666136
	end_time_unix_nano: 1737371954188794618
	attributes {
	  key: "url.path"
	  value {
	    string_value: "/dynamicconfig/flipSending"
	  }
	}
	attributes {
	  key: "server.address"
	  value {
	    string_value: "localhost"
	  }
	}
	attributes {
	  key: "client.address"
	  value {
	    string_value: "172.18.0.1"
	  }
	}
	attributes {
	  key: "network.peer.address"
	  value {
	    string_value: "172.18.0.1"
	  }
	}
	attributes {
	  key: "network.peer.port"
	  value {
	    int_value: 53180
	  }
	}
	attributes {
	  key: "http.route"
	  value {
	    string_value: "/dynamicconfig/flipSending"
	  }
	}
	attributes {
	  key: "http.request.method"
	  value {
	    string_value: "GET"
	  }
	}
	attributes {
	  key: "http.response.status_code"
	  value {
	    int_value: 200
	  }
	}
	attributes {
	  key: "user_agent.original"
	  value {
	    string_value: "okhttp/4.12.0"
	  }
	}
	attributes {
	  key: "thread.id"
	  value {
	    int_value: 36
	  }
	}
	attributes {
	  key: "network.protocol.version"
	  value {
	    string_value: "1.1"
	  }
	}
	attributes {
	  key: "server.port"
	  value {
	    int_value: 32774
	  }
	}
	attributes {
	  key: "url.scheme"
	  value {
	    string_value: "http"
	  }
	}
	attributes {
	  key: "thread.name"
	  value {
	    string_value: "http-nio-8080-exec-3"
	  }
	}
	status {
	}
	]
```
</details>

## Additional context

Another flake of the same test with a different assertion failure:


<details>
<summary>Click to expand</summary>

```

java.lang.AssertionError: 
Expected size: 2 but was: 0 in:
[]
	at app//com.example.javaagent.smoketest.DynamicConfigSmokeTest.flipSending(DynamicConfigSmokeTest.java:66)
	at java.base@21.0.5/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base@21.0.5/java.lang.reflect.Method.invoke(Method.java:586)
	at app//org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:767)
	at app//org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:156)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:147)
	at app//org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:86)
	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(InterceptingExecutableInvoker.java:103)
	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at app//org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
	at app//org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:86)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$8(TestMethodTestDescriptor.java:217)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:213)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:138)
	at app//org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:68)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:156)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:146)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:144)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:143)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:100)
	at java.base@21.0.5/java.util.ArrayList.forEach(ArrayList.java:1596)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:160)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:146)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:144)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:143)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:100)
	at java.base@21.0.5/java.util.ArrayList.forEach(ArrayList.java:1596)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:160)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:146)
	at app//org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:144)
	at app//org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:143)
	at app//org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:100)
	at app//org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at app//org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at app//org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at app//org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at app//org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:124)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:94)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:63)
	at java.base@21.0.5/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base@21.0.5/java.lang.reflect.Method.invoke(Method.java:586)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:92)
	at jdk.proxy1/jdk.proxy1.$Proxy4.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:200)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:132)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:103)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:63)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:121)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

--- a/smoke-tests/src/test/java/com/example/javaagent/smoketest/DynamicConfigSmokeTest.java
+++ b/smoke-tests/src/test/java/com/example/javaagent/smoketest/DynamicConfigSmokeTest.java
@@ -20,9 +20,9 @@ package com.example.javaagent.smoketest;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-import com.google.protobuf.ByteString;
 import io.opentelemetry.proto.collector.trace.v1.ExportTraceServiceRequest;
 import io.opentelemetry.proto.trace.v1.Span;
+import java.io.IOException;
 import java.util.List;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.AfterEach;
@@ -54,32 +54,25 @@ class DynamicConfigSmokeTest extends TestAppSmokeTest {
   @AfterEach
   public void endTest() throws InterruptedException {
     doRequest(getUrl("/dynamicconfig/reset"), okResponseBody("reset"));
-    Thread.sleep(1500L); // give the reset time to be applied
   }
 
   @Test
-  public void flipSending() throws InterruptedException {
+  public void flipSending() throws InterruptedException, IOException {
+    doRequest(getUrl("/health"), okResponseBody("Alive!"));
     doRequest(getUrl("/dynamicconfig/flipSending"), okResponseBody("stopped"));
+
+    // the first flip-sending request should not be part of the exported traces anymore
     List<ExportTraceServiceRequest> traces = waitForTraces();
     List<Span> spans = getSpans(traces).toList();
-    assertThat(spans)
-        .hasSize(2)
-        .extracting("name")
-        .containsOnly("GET /dynamicconfig/flipSending", "DynamicConfigController.flipSending");
-    ByteString firstTraceID = spans.get(0).getTraceId();
+    assertThat(spans).hasSize(1).extracting("name").containsOnly("GET /health");
 
-    Thread.sleep(2000L); // give the flip time to be applied
+    clearBackend();
 
+    doRequest(getUrl("/health"), okResponseBody("Alive!"));
     doRequest(getUrl("/dynamicconfig/flipSending"), okResponseBody("restarted"));
+    // During /health, the sending should still have been disabled, /flipSending should be recorded
     traces = waitForTraces();
-    spans = getSpans(traces).dropWhile(span -> span.getTraceId().equals(firstTraceID)).toList();
-    assertThat(spans).hasSize(0);
-
-    Thread.sleep(2000L); // give the flip time to be applied
-
-    doRequest(getUrl("/dynamicconfig/flipSending"), okResponseBody("stopped"));
-    traces = waitForTraces();
-    spans = getSpans(traces).dropWhile(span -> span.getTraceId().equals(firstTraceID)).toList();
+    spans = getSpans(traces).toList();
     assertThat(spans)
         .hasSize(2)
         .extracting("name")

--- a/smoke-tests/test-app/src/main/java/co/elastic/otel/test/DynamicConfigController.java
+++ b/smoke-tests/test-app/src/main/java/co/elastic/otel/test/DynamicConfigController.java
@@ -31,18 +31,33 @@ public class DynamicConfigController {
   @GetMapping("/flipSending")
   public synchronized String flipSending() {
     String old = System.getProperty(DISABLE_SEND_OPTION, "");
+    String result;
     if (old.isEmpty()) {
       System.setProperty(DISABLE_SEND_OPTION, "true");
-      return "stopped";
+      result = "stopped";
     } else {
       System.setProperty(DISABLE_SEND_OPTION, "");
-      return "restarted";
+      result = "restarted";
+    }
+    waitForSystemPropertyChangeEffective();
+    return result;
+  }
+
+  private static void waitForSystemPropertyChangeEffective() {
+    try {
+      Thread.sleep(1000L);
+    } catch (InterruptedException e) {
+      throw new IllegalStateException(e);
     }
   }
 
   @RequestMapping("/reset")
   public synchronized String reset() {
-    System.setProperty(DISABLE_SEND_OPTION, "");
+    String old = System.getProperty(DISABLE_SEND_OPTION, "");
+    if (!old.isEmpty()) {
+      System.setProperty(DISABLE_SEND_OPTION, "");
+      waitForSystemPropertyChangeEffective();
+    }
     return "reset";
   }
 }

